internal/pipeline/types_test.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package pipeline_test
+
+import (
+	"testing"
+
+	. "github.com/GoogleContainerTools/kpt/internal/pipeline"
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+func parseFromString(input string) (Pipeline, error) {
+	var p Pipeline
+	err := yaml.Unmarshal([]byte(input), &p)
+	if err != nil {
+		return p, err
+	}
+	return p, nil
+}
+
+func TestPipelineParse(t *testing.T) {
+	input := `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+  - ./base
+  - ./*
+
+generators:
+  - image: gcr.io/kpt-functions/generate-folders
+transformers:
+  - image: patch-strategic-merge
+    configPath: ./patch.yaml
+  - image: gcr.io/kpt-functions/set-annotation
+    configMap:
+      environment: dev
+
+validators:
+  - image: gcr.io/kpt-functions/policy-controller-validate
+`
+
+	actual, err := parseFromString(input)
+	assert.NoError(t, err)
+	expected := Pipeline{
```

Comment:
Just a thought... I am realizing that we will probably need a pipeline.Builder() because when we will be reading packages without pipeline definitions, we will have to build pipeline with zero configuration etc and we can add to the pipeline. So something like:

b := pipeline.Builder().WithSources([]string).WithGenerators([]Generator)...
pipeline, err := b.Build()

Reply:
Good idea. Will consider it when I work on pipeline IO